src/jdk.compiler/share/classes/com/sun/tools/javac/code/LintMapper.java

```diff
@@ -0,0 +1,329 @@
+
+        final LintRange rootRange;                              // the root LintRange (covering the entire source file)
+        final List<Span> unmappedDecls = new ArrayList<>();     // unmapped top-level declarations awaiting attribution
+
+        // After parsing: Add top-level declarations to our "unmappedDecls" list
+        FileInfo(Lint rootLint, JCCompilationUnit tree) {
+            rootRange = new LintRange(rootLint);
+            tree.defs.stream()
+              .filter(this::isTopLevelDecl)
+              .map(decl -> new Span(decl, tree.endPositions))
+              .forEach(unmappedDecls::add);
+        }
+
+        // After attribution: Discard the span from "unmappedDecls" and populate the declaration's subtree under "rootRange"
+        void afterAttr(JCTree tree, EndPosTable endPositions) {
+            for (Iterator<Span> i = unmappedDecls.iterator(); i.hasNext(); ) {
+                if (i.next().contains(tree.pos())) {
+                    rootRange.populateSubtree(tree, endPositions);
+                    i.remove();
+                    return;
+                }
+            }
+            throw new AssertionError("top-level declaration not found");
+        }
+
+        // Find the most specific Lint configuration applying to the given position, unless the position has not been mapped yet
+        Optional<Lint> lintAt(DiagnosticPosition pos) {
+            boolean mapped = unmappedDecls.stream().noneMatch(span -> span.contains(pos));
+            return mapped ? Optional.of(rootRange.bestMatch(pos).lint) : Optional.empty();
+        }
+
+        boolean isTopLevelDecl(JCTree tree) {
+            return tree.getTag() == Tag.MODULEDEF
+                || tree.getTag() == Tag.PACKAGEDEF
+                || tree.getTag() == Tag.CLASSDEF;
+        }
+    }
+
+// Span
+
+    /**
+     * A lexical range.
+     */
+    private record Span(int startPos, int endPos) {
+
+        static final Span MAXIMAL = new Span(Integer.MIN_VALUE, Integer.MAX_VALUE);
+
+        Span(JCTree tree, EndPosTable endPositions) {
+            this(TreeInfo.getStartPos(tree), TreeInfo.getEndPos(tree, endPositions));
+        }
+
+        boolean contains(DiagnosticPosition pos) {
+            int offset = pos.getLintPosition();
+            return offset == startPos || (offset > startPos && offset < endPos);
+        }
+
+        boolean contains(Span that) {
+            return this.startPos <= that.startPos && this.endPos >= that.endPos;
+        }
+    }
+
+// LintRange
+
+    /**
+     * A tree of nested lexical ranges and the {@link Lint} configurations that apply therein.
+     */
+    private record LintRange(
+        Span span,                                      // declaration's lexical range
+        Lint lint,                                      // the Lint configuration that applies at this declaration
+        Symbol symbol,                                  // declaration symbol (for debug purposes only; null for root)
```

Comment:
nit: should we keep this field in the final version? we can add it while debugging if needed

Reply:
> nit: should we keep this field in the final version? we can add it while debugging if needed

I can remove it. Besides, this symbol is always the same object as the one that the `Lint` instance was created with, so if debugging were a concern, a better place to store it would be with the `Lint` instance anyway.

Reply:
Unused field removed in 1f1247bc5cf.